sycl/source/detail/graph_impl.hpp

```diff
@@ -154,16 +154,32 @@ class node_impl {
         MCGType(Other.MCGType), MNodeType(Other.MNodeType),
         MCommandGroup(Other.getCGCopy()), MSubGraphImpl(Other.MSubGraphImpl) {}
 
-  /// Checks if this node has a given requirement.
-  /// @param Requirement Requirement to lookup.
-  /// @return True if \p Requirement is present in node, false otherwise.
-  bool hasRequirement(sycl::detail::AccessorImplHost *IncomingReq) {
+  /// Checks if this node should be a dependency of another node based on
+  /// accessor requirements. This is calculated using access modes if a
+  /// requirement to the same buffer is found inside this node.
+  /// @param IncomingReq Incoming requirement.
+  /// @return True if a dependency is needed, false if not.
+  bool hasRequirementDependency(sycl::detail::AccessorImplHost *IncomingReq) {
+    access_mode InMode = IncomingReq->MAccessMode;
+    // If the incoming requirement is write only or discard_*, we don't create
+    // a dependency
+    if (InMode != access_mode::read && InMode != access_mode::read_write &&
```

Comment:
Does it make sense to switch this to check explicitly for dependency-free access modes? This could prevent a missed dependency in case new access modes are added in the future.